src/nestegg.c

```
@@ -3458,3 +3458,210 @@ int nestegg_sniff_mkv(unsigned char const* buffer, size_t length)
+static int
+ne_sum_block_or_group(nestegg * ctx, uint64_t id, uint64_t size, uint64_t * framesOut)
+{
+  int r;
+
+  if (id == ID_SIMPLE_BLOCK) {
+    uint64_t frames;
+    frames = 0;
+    r = ne_read_block_lacing(ctx, size, &frames);
+    if (r != 1)
+      return r;
+    *framesOut += frames;
+    return 1;
+  }
+
+  if (id == ID_BLOCK_GROUP) {
+    int64_t group_end;
+    group_end = ne_io_tell(ctx->io) + (int64_t) size;
+    while (ne_io_tell(ctx->io) < group_end) {
+      uint64_t gid, gsize;
+      r = ne_read_element(ctx, &gid, &gsize);
+      if (r != 1)
+        return r;
+
+      if (gid == ID_BLOCK) {
+        uint64_t frames;
+        frames = 0;
+        r = ne_read_block_lacing(ctx, gsize, &frames);
+        if (r != 1)
+          return r;
+        *framesOut += frames;
+      } else {
+        r = ne_io_read_skip(ctx->io, gsize);
+        if (r != 1)
+          return r;
+      }
+    }
+    return 1;
+  }
+
+  /* Not a block-like element: skip its payload. */
+  r = ne_io_read_skip(ctx->io, size);
+  if (r != 1)
+    return r;
+
+  return 1;
+}
+
+/* Read ONE Cluster and return the sum of frames of ALL SimpleBlock/Block in it.
+   Returns 1 on success (Cluster found), 0 on clean EOS before any Cluster,
+   <0 on error. */
+static int
+ne_read_cluster_frames_count(nestegg * ctx, uint64_t * framesOut)
+{
+  int r;
+  uint64_t id, size;
+  int64_t cluster_end;
+  uint64_t totalFrames;
+
+  assert(ctx->ancestor == NULL);
+
+  /* Find the next Cluster at the top level. */
+  for (;;) {
+    r = ne_read_element(ctx, &id, &size);
+    if (r == 0) {
+      return 0; /* EOS before any Cluster */
+    }
+    if (r != 1) {
+      return r; /* error */
+    }
+
+    if (id != ID_CLUSTER) {
+      /* Not a Cluster: consume and keep scanning. */
+      r = ne_io_read_skip(ctx->io, size);
+      if (r != 1)
+        return r;
+      continue;
+    }
+
+    /* Enter Cluster body by range. */
+    cluster_end = ne_io_tell(ctx->io) + (int64_t) size;
```

Note that Clusters may have an "unknown" size, we should handle this somehow.  The rest of the code has generally tried to avoid needing to know where a Cluster ends to avoid this issue.  It's probably uncommon enough that we can simply check for "unknown" size and error out, since I don't believe the "unknown" size feature is used except for Segments.